File: src/generics/generics.cs
using System.Reflection;
using Xunit;

namespace ZigForDotnetDevs;

public class GenericType<T>
{
    public T GenericField;
}

public class Generics
{
    public string GenericFunction<T>(T genericParameter)
    {
        return genericParameter.ToString();
    }

    [Fact]
    public void GenericTypeTest()
    {
        var instance = new GenericType<int>();
        instance.GenericField = 123;
        Assert.Equal(123, instance.GenericField);
    }

    [Fact]
    public void GenericFunctionTest()
    {
        var result = GenericFunction<int>(123);
        Assert.Equal("123", result);
    }
}

File: src/generics/generics.zig
const std = @import("std");

pub fn GenericType(T: type) type {
    return struct {
        generic_field: T,
    };
}

test GenericType {
    const instance = GenericType(i32){ .generic_field = 123 };
    try std.testing.expectEqual(@as(i32, 123), instance.generic_field);
}

pub fn genericFunction(T: type, generic_parameter: T, allocator: std.mem.Allocator) ![]const u8 {
    return std.fmt.allocPrint(allocator, "{d}", .{generic_parameter});
}

test genericFunction {
    const result = try genericFunction(i32, 123, std.testing.allocator);
    try std.testing.expectEqualSlices(u8, "123", result);
}
